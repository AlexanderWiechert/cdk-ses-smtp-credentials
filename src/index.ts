export * from "./ses-smtp-credentials";
