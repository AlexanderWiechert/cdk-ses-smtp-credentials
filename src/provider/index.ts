export * from "./calculate-ses-smtp-password";
export * from "./credentials";
export * from "./credentials-handler.lambda";
export * from "./credentials-handler-function";
export * from "./credentials-provider";
