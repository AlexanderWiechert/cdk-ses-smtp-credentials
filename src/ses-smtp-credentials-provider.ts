import * as path from "path";
import { IUser, PolicyStatement } from "@aws-cdk/aws-iam";
import { Runtime } from "@aws-cdk/aws-lambda";
import { NodejsFunction, NodejsFunctionProps } from "@aws-cdk/aws-lambda-nodejs";
import { RetentionDays } from "@aws-cdk/aws-logs";
import { ISecret } from "@aws-cdk/aws-secretsmanager";
import { Construct, Duration } from "@aws-cdk/core";
import { Provider } from "@aws-cdk/custom-resources";

interface SesSmtpCredentialsProviderProps {
  readonly user: IUser;
  readonly secret: ISecret;
}

export class SesSmtpCredentialsProvider extends Construct {
  public readonly serviceToken: string;

  constructor(scope: Construct, id: string, props: SesSmtpCredentialsProviderProps) {
    super(scope, id);

    const { secret, user } = props;

    // https://docs.aws.amazon.com/cdk/api/latest/docs/aws-lambda-nodejs-readme.html
    const onEvent = new NodejsFunction(this, "ses-smtp-credentials-handler", {
      runtime: Runtime.NODEJS_14_X,
      entry: path.join(__dirname, "provider", "handler.ts"),
      handler: "onEvent",
      timeout: Duration.minutes(1),
      initialPolicy: [
        new PolicyStatement({
          resources: [user.userArn],
          actions: ["iam:CreateAccessKey", "iam:DeleteAccessKey"],
        }),
        new PolicyStatement({
          resources: [secret.secretArn],
          actions: ["iam:putSecretValue"],
        }),
      ],
    } as NodejsFunctionProps);

    const provider = new Provider(this, "ses-smtp-credentials-provider", {
      onEventHandler: onEvent,
      logRetention: RetentionDays.ONE_DAY, // default is INFINITE
    });

    this.serviceToken = provider.serviceToken;
  }
}
