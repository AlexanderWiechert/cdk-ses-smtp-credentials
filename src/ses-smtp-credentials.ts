import { Annotations, CustomResource, CustomResourceProps, Stack } from "aws-cdk-lib";
import * as iam from "aws-cdk-lib/aws-iam";
import * as secretsmanager from "aws-cdk-lib/aws-secretsmanager";
import { Construct } from "constructs";
import { CredentialsProvider } from "./provider/credentials-provider";
import { SMTP_REGIONS } from "./provider/ses-smtp-regions";

export interface SesSmtpCredentialsProps {
  /**
   * The user for which to create an AWS Access Key and to generate the smtp password. If omitted a user will be created.
   */
  readonly user?: iam.IUser;
  /**
   * Optional, a username to create a new user if no existing user is given.
   */
  readonly userName?: string;
  /**
   * Optional, an SecretsManager secret to write the AWS SES Smtp credentials to.
   */
  readonly secret?: secretsmanager.ISecret;
}

/**
 * This construct creates an access key for the given user and stores the generated SMTP credentials inside a secret.
 *
 * Attaches an inline policy to the user allowing to send emails
 *
 * ```typescript
 * const user = User.fromUserName("ses-user-example");
 * const credentials = new SesSmtpCredentials(this, 'SmtpCredentials', {
 *     user: user,
 * });
 * // smtpCredentials.secret contains json value {username: "<the generated access key id>", password: "<the calculated ses smtp password>"}
 * ```
 */
export class SesSmtpCredentials extends Construct {
  /**
   * The secret that contains the calculated AWS SES Smtp Credentials.
   *
   * ```typescript
   * import { aws_ecs } from "aws-cdk-lib";
   *
   * const containerDefinitionOptions: aws_ecs.ContainerDefinitionOptions = {
   *     // ...
   *     secrets: {
   *         MAIL_USERNAME: aws_ecs.Secret.fromSecretsManager(smtpCredentials.secret, "username"),
   *         MAIL_PASSWORD: aws_ecs.Secret.fromSecretsManager(smtpCredentials.secret, "password"),
   *     }
   * }
   * ```
   */
  public readonly secret: secretsmanager.ISecret;

  public constructor(scope: Construct, id: string, props: SesSmtpCredentialsProps) {
    super(scope, id);

    const region = Stack.of(this).region;
    if (!SMTP_REGIONS.includes(region)) {
      Annotations.of(this).addWarning(
        `AWS SES Smtp Endpoint is not available in region ${region}\n see https://docs.aws.amazon.com/general/latest/gr/ses.html`,
      );
    }

    const user =
      props.user ??
      new iam.User(this, "User", {
        userName: props.userName,
      });

    const policy = new iam.Policy(this, "Policy", {
      statements: [
        new iam.PolicyStatement({
          // https://github.com/awsdocs/amazon-ses-developer-guide/blob/master/doc-source/sending-authorization-policies.md
          effect: iam.Effect.ALLOW,
          actions: ["ses:SendRawEmail"],
          resources: ["*"],
        }),
      ],
    });
    user.attachInlinePolicy(policy);

    this.secret =
      props.secret ||
      new secretsmanager.Secret(this, "Secret", {
        description: `SES Smtp Credentials (username, password) for ${user.userName}`,
      });

    const provider = CredentialsProvider.getOrCreate(this);
    provider.grant(user, this.secret);

    const customResource = new CustomResource(this, "Lambda", {
      serviceToken: provider.serviceToken,
      properties: {
        UserName: user.userName,
        SecretId: this.secret.secretArn,
      },
    } as CustomResourceProps);
    customResource.node.addDependency(user);
    customResource.node.addDependency(this.secret);
  }
}
