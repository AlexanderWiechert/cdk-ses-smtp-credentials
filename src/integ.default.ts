import { App, Environment, Stack } from "aws-cdk-lib";
import { User } from "aws-cdk-lib/aws-iam";
import { SesSmtpCredentials } from "./ses-smtp-credentials";

// export CDK_DEFAULT_REGION=us-east-1
// export CDK_DEFAULT_ACCOUNT=123456789012
// projen bundle:provider/credentials-handler.lambda
// cdk deploy --app 'npx ts-node -P tsconfig.json --prefer-ts-exts ./src/integ.default.ts'

export const app = new App();

const env: Environment = {
  region: process.env.CDK_DEFAULT_REGION,
  account: process.env.CDK_DEFAULT_ACCOUNT,
};

export const stack = new Stack(app, "SesSmtpCredentialsIntegrationTest", { env: env });

const user = new User(stack, "User", { userName: "SesTestUser" });

new SesSmtpCredentials(stack, "Credentials", { user: user });
