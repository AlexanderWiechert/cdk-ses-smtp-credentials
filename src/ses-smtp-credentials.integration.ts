import { User } from "@aws-cdk/aws-iam";
import { App, Stack } from "@aws-cdk/core";
import { SesSmtpCredentials } from "./ses-smtp-credentials";

const app = new App();
const stack = new Stack(app, "SesSmtpCredentialsStack", {});

const user = new User(stack, "User", {});
new SesSmtpCredentials(stack, "Credentials", {
  user: user,
});

app.synth();
