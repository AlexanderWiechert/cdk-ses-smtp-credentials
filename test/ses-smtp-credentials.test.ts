import { App, Aspects, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { User } from "aws-cdk-lib/aws-iam";
import { AwsSolutionsChecks } from "cdk-nag";
import { SesSmtpCredentials } from "../src";

jest.mock("../src/ses-smtp-credentials-provider.ts");

describe("SesSmtpCredentials", () => {
  it("Should match snapshot", () => {
    // Given
    const app = new App();
    const stack = new Stack(app);
    const user = new User(stack, "User", {
      userName: "Micha",
    });

    // Then
    new SesSmtpCredentials(stack, "SesSmtpCredentials", {
      user: user,
    });
    const template = Template.fromStack(stack);

    // Then
    expect(template).toMatchSnapshot();
  });

  it("Should comply to best practices", () => {
    // Given
    const app = new App();
    const stack = new Stack(app);
    const user = new User(stack, "User", {
      userName: "Micha",
    });

    // Then
    new SesSmtpCredentials(stack, "SesSmtpCredentials", {
      user: user,
    });

    // Then
    Aspects.of(app).add(new AwsSolutionsChecks({ verbose: true }));
  });
});
