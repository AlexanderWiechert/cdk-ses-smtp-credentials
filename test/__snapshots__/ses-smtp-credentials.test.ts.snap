// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SesSmtpCredentials Should match snapshot 1`] = `
Object {
  "Resources": Object {
    "SesSmtpCredentialsLambda69DED5B2": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "UserName": "Micha",
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
